### DISCUSSION TITLE
SOPS decrypt not working in Flux with age key

### DISCUSSION THREAD

**Discussion by archana-punia** [fluxcd/flux2#3210] [Q&A] [answered]
I have created everything like mentioned in official release but still getting below error

{"level":"error","ts":"2022-10-14T01:10:11.167Z","logger":"controller.kustomization","msg":"Reconciliation failed after 39.111201ms, next try in 10m0s","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"secrets","namespace":"flux","revision":"main/20bc668c0ac20b59824abbec0fc317d374490b2a","error":"Secret/flux/azure-secret-sops conflict, error: data values must be of type string

my version details are:
flux: v0.31.3
helm-controller: v0.22.1
kustomize-controller: v0.26.2
notification-controller: v0.24.0
source-controller: v0.25.9

 ===========================================
> secret kustomization
```
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: secrets
  namespace: flux
spec:
  interval: 10m0s
  sourceRef:
    kind: GitRepository
    name: flux
    namespace: flux
  path: ./clusters/staging/flux/secrets
  prune: true
  decryption:
    provider: sops
    secretRef:
      name: sops-age
```

> encrypted secret
```
apiVersion: v1
stringData:
    azurestorageaccountkey: ENC[AES256_GCM,data:qP/r7PCS3AfWG8YNUL2wCLx8uScyVC7uJD3nP1zZ8oVCWATcu1qN2u8wuO6OkMUS8LS+wWqMi6aX9wNWAjp/UqOi4TxKKTbEKL95L43WHnb4xzKXNINPrgbPTO/e9DIjy9YJNhynhRZQOBEHV86txKUYNctiIcRp,iv:SDLvRZ9wVLB9W161Ow31n85iNZ1fpFv3G0NPzX28yg4=,tag:uX9jSa0TUpMGH43D5g6KZA==,type:str]
    azurestorageaccountname: ENC[AES256_GCM,data:oQ66HdDwViywJQ1ZVMldgXcc3y4=,iv:PKDszsf+ADmRCM1rvp6Vxyct5IRskFwk9NzZrqOowkQ=,tag:7FYEjQH4VjWc5QVVicSw9g==,type:str]
kind: Secret
metadata:
    annotations:
        meta.helm.sh/release-name: test
        meta.helm.sh/release-namespace: test
    creationTimestamp: "2022-08-04T18:46:35Z"
    labels:
        app.kubernetes.io/managed-by: Helm
        helm.toolkit.fluxcd.io/name: test
        helm.toolkit.fluxcd.io/namespace: test
    name: azure-secret-sops
    namespace: flux
    resourceVersion: "45187684"
    uid: 90ece448-ace6-4d7d-a6c5-e52851111dc9
type: Opaque
sops:
    kms: []
    gcp_kms: []
    azure_kv: []
    hc_vault: []
    age:
        - recipient: age1xfaapdzdfvhknhjk7jq68sxruxhxjfe9e9hvssv345kjkn5p0ql2686j
          enc: |
            -----BEGIN AGE ENCRYPTED FILE-----
            YWdlLWVuY3J5cHRpb24ub3JnL3YxCi0+IFgyNTUxOSBDNGpxb2JGZUp4TjlFSzdT
            ZENqcVVEbzJqUVZlQkVRRmdybmk4TVNCa2c0CnA3VGZFTyszRnFBN1d2MWkrZy8w
            bmFzT0lBdkhmS2g5Wk10Z2haMCtSaVUKLS0tIFlpZmppbFhMbGF2eGJLbzNSbEY4
            ekQ1MVRyc0NNdEdhYVIrWUlZZW5tMXcKAimVUt7HWcs7vH6IyXPjPJGomM+PyW6/
            CgjZcmAxGNHUHm3+PP6Cco1BCqimI34MHJgjyhODvBUxOpJ6AZHXsA==
            -----END AGE ENCRYPTED FILE-----
    lastmodified: "2022-10-13T21:38:46Z"
    mac: ENC[AES256_GCM,data:I3JloYz4CXZAXLCLvLMFbWMbcGpQkb2mcVCap/hBWPlTFRBlQw5+EFp7oPj0HB5DvOoNLmwgdQnJNNGegb6dnTVgcybGQRv/8yubsuy3F6IQebBcIg7lV+EM3B0eY+gcq8KWbiDDtEKd6R3By8YHt0Y2McFaxODs3f2oBG9WkJk=,iv:utTzszhHLCbYyFWzEuSsCjhMCr3RwjfMvR/KSZmSfBw=,tag:RbEjWU0k1jn6ZKXxQI3COg==,type:str]
    pgp: []
    encrypted_regex: ^(data|stringData)$
    version: 3.7.3
```

================================================================

can anyone help out here?

**Comment by stefanprodan** [comment, marked as answer]
You can't have uids and helm annotations in Git. Please remove all the read-only fields from the secret.

**Reply by archana-punia** [reply]
You are right. Once I removed these, it created the secret. But now I am getting another issue, when it applies the secret to the cluster, the values are getting changed.

When I checked within cluster, it like base64 encoded version of already encoded values in secret. Like actual value>base64>base64, thats what m seeing now. so, my applications now sees a base64 version of actual value and throws error.

When I manually decode the same secret yaml using sops, I see the correct value but somehow Flux is not doing the decryption correctly while applying it to the cluster.

**Reply by stefanprodan** [reply]
Start fresh with a plain text secret:

```yaml
apiVersion: v1
kind: Secret
metadata:
    name: azure-secret-sops
    namespace: flux
stringData:
    azurestorageaccountkey:  <PLAIN-TEXT-HERE>
    azurestorageaccountname: <PLAIN-TEXT-HERE>
```
Save this to disk as secret.yaml and encrypt it-place with SOPS.

**Reply by archana-punia** [reply]
creating it from scratch. Working good now. Thank you so much for your help. 👍

**Reply by spuny** [reply]
Yes stringData works but this does not work with data only

this one will result in error:
`Secret/default/basic-auth validation error: error decoding from json: illegal base64 data at input byte 3`

file:
```
apiVersion: v1
data:
  password: Y2hhbmdlbWU=
  user: YWRtaW4=
kind: Secret
metadata:
  creationTimestamp: null
  name: basic-auth
  namespace: default
```
but this one will be applied without any problem:
```
apiVersion: v1
kind: Secret
metadata:
    name: basic-auth-string-data
    namespace: default
stringData:
    password: changeme
    user: admin
```

Yes i've tried to create it with only plain text in both versions but only the one with stringData seems to work just fine.
EDIT: just in case, the `basic-auth-string-data` works even encoded. No idea why `data:` does not

EDIT2: Ok figured it out, `stringData` just does not care if it is in correct format. In this case it just base64 encoded the encrypted values:
```
apiVersion: v1
data:
  password: RU5DW0FFUzI1Nl9HQ00sZGF0YTpPVFk0MFlic0tROEEvdlFjLGl2OjdMb1lYM2sweFhGNDQwb3d5d0FSM0pUblhLNnpGMG82VkpRRkk4bTdFck09LHRhZzowYm0yQWNDaTVDdWRkdk5xWmp5Z0tRPT0sdHlwZTpzdHJd
  user: RU5DW0FFUzI1Nl9HQ00sZGF0YTpOSXNScUdEb0xCTT0saXY6NnBDbFc5NGNjaW56ZjI2WFNHOFJuRjBwMmJRQnJ0bnd4Zmo0RUsrY3JETT0sdGFnOi91VnVOUmZ3RUxxU1U1clJmRVFnQUE9PSx0eXBlOnN0cl0=
kind: Secret
metadata:
  creationTimestamp: "2023-06-23T09:51:47Z"
  labels:
    kustomize.toolkit.fluxcd.io/name: apps
    kustomize.toolkit.fluxcd.io/namespace: flux-system
  name: basic-auth-string-data
  namespace: default
  resourceVersion: "169642"
  uid: a08094bd-7058-43cb-8616-626efe5d368d
type: Opaque


❯ echo "RU5DW0FFUzI1Nl9HQ00sZGF0YTpPVFk0MFlic0tROEEvdlFjLGl2OjdMb1lYM2sweFhGNDQwb3d5d0FSM0pUblhLNnpGMG82VkpRRkk4bTdFck09LHRhZzowYm0yQWNDaTVDdWRkdk5xWmp5Z0tRPT0sdHlwZTpzdHJd" |base64 -d
ENC[AES256_GCM,data:OTY40YbsKQ8A/vQc,iv:7LoYX3k0xXF440owywAR3JTnXK6zF0o6VJQFI8m7ErM=,tag:0bm2AcCi5CuddvNqZjygKQ==,type:str]%
```
